akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMergeSpec.scala

```diff
@@ -96,5 +96,30 @@ class FlowMergeSpec extends BaseTwoStreamsSetup {
       up1.expectSubscription().expectCancellation()
       up2.expectSubscription().expectCancellation()
     }
+
+    "not try to grab from a closed input previously enqueued" in assertAllStagesStopped {
+      // Coverage for #21138
+      val up1 = TestPublisher.probe[Int]()
+      val up2 = TestPublisher.probe[Int]()
+      val down = TestSubscriber.probe[Int]()
+
+      Source.fromPublisher(up1)
+        .merge(Source.fromPublisher(up2), eagerComplete = true)
+        .to(Sink.fromSubscriber(down))
+        .run
+
+      up1.ensureSubscription()
+      up2.ensureSubscription()
+      down.ensureSubscription()
+
+      up1.expectRequest()
+      up2.expectRequest()
+      up1.sendNext(7)
+      up2.sendNext(8)
+      up2.sendComplete()
+      down.request(1)
```

Comment:
You can't really enforce non-racyness here :( In test where it mattered I had to add a sleep. Since this test will not fail no matter how the race works out, I guess it is fine to leave it as is, maybe add a comment.